src/apps/js_generic/js_generic_base.cpp

```diff
@@ -399,35 +366,37 @@ namespace ccfapp
                 "conversion of body).");
               return;
             }
-            std::string str(cstr);
-            JS_FreeCString(ctx, cstr);
 
-            response_body = std::vector<uint8_t>(str.begin(), str.end());
+            response_body = std::vector<uint8_t>(str->begin(), str->end());
           }
           endpoint_ctx.rpc_ctx->set_response_body(std::move(response_body));
         }
       }
       // Response headers
       {
-        auto response_headers_js = ctx(JS_GetPropertyStr(ctx, val, "headers"));
+        auto response_headers_js = val["headers"];
         if (JS_IsObject(response_headers_js))
         {
           uint32_t prop_count = 0;
-          JSPropertyEnum* props = nullptr;
+          JSPropertyEnum* prop_enum = nullptr;
           JS_GetOwnPropertyNames(
             ctx,
-            &props,
+            &prop_enum,
             &prop_count,
             response_headers_js,
             JS_GPN_STRING_MASK | JS_GPN_ENUM_ONLY);
           for (size_t i = 0; i < prop_count; i++)
           {
-            auto prop_name = props[i].atom;
-            auto prop_name_cstr = ctx(JS_AtomToCString(ctx, prop_name));
-            auto prop_val =
-              ctx(JS_GetProperty(ctx, response_headers_js, prop_name));
-            auto prop_val_cstr = JS_ToCString(ctx, prop_val);
-            if (!prop_val_cstr)
+            auto prop_name = prop_enum[i].atom;
+            auto prop_name_str = ctx.to_str(prop_name);
+            if (!prop_name_str)
```

Comment:
I think for symmetry this should set_error like line 401 with a message like "Invalid endpoint function return value (header type).".